Cast Exception when editing Narratives Time in Full Debrief

Steps to reproduce:
1. Open narrative.rep
2. Edit time of any narrative entry.

> !STACK 0
> java.lang.IndexOutOfBoundsException: Cannot get at -1 on list of size 38
> 	at ca.odell.glazedlists.TransformedList.get(TransformedList.java:105)
> 	at org.eclipse.nebula.widgets.nattable.data.ListDataProvider.getRowObject(ListDataProvider.java:62)
> 	at org.mwc.cmap.NarrativeViewer2.NatNarrativeViewer.getSelection(NatNarrativeViewer.java:492)
> 	at org.mwc.cmap.NarrativeViewer2.NatNarrativeViewer.setDTG(NatNarrativeViewer.java:581)
> 	at org.mwc.cmap.NarrativeViewer2.NATViewerView$27.run(NATViewerView.java:1330)
> 	at org.eclipse.swt.widgets.RunnableLock.run(RunnableLock.java:40)
> 	at org.eclipse.swt.widgets.Synchronizer.runAsyncMessages(Synchronizer.java:185)
> 	at org.eclipse.swt.widgets.Display.runAsyncMessages(Display.java:5113)
> 	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:4666)
> 	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1173)
> 	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
> 	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1062)
> 	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155)
> 	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:635)
> 	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
> 	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:559)
> 	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:150)
> 	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:155)
> 	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)
> 	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:137)
> 	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:107)
> 	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:400)
> 	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
> 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> 	at java.lang.reflect.Method.invoke(Method.java:498)
> 	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:660)
> 	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:597)
> 	at org.eclipse.equinox.launcher.Main.run(Main.java:1468)
> 	at org.eclipse.equinox.launcher.Main.main(Main.java:1441)
> 
> !ENTRY org.mwc.cmap.core 4 0 2019-09-26 12:53:47.228
> !MESSAGE Whilst adding new action to history buffer
> !STACK 0
> org.eclipse.core.commands.ExecutionException: While executing the operation, an exception occurred
> 	at org.eclipse.core.commands.operations.DefaultOperationHistory.execute(DefaultOperationHistory.java:503)
> 	at org.mwc.cmap.core.CorePlugin.run(CorePlugin.java:823)
> 	at org.mwc.cmap.core.property_support.EditableWrapper.setPropertyValue(EditableWrapper.java:772)
> 	at org.eclipse.ui.views.properties.PropertySheetEntry.valueChanged(PropertySheetEntry.java:751)
> 	at org.eclipse.ui.views.properties.PropertySheetEntry.valueChanged(PropertySheetEntry.java:756)
> 	at org.eclipse.ui.views.properties.PropertySheetEntry.setValue(PropertySheetEntry.java:691)
> 	at org.eclipse.ui.views.properties.PropertySheetEntry.applyEditorValue(PropertySheetEntry.java:153)
> 	at org.eclipse.ui.views.properties.PropertySheetEntry$1.applyEditorValue(PropertySheetEntry.java:114)
> 	at org.eclipse.jface.viewers.CellEditor$1.run(CellEditor.java:335)
> 	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:45)
> 	at org.eclipse.ui.internal.JFaceUtil.lambda$0(JFaceUtil.java:47)
> 	at org.eclipse.jface.util.SafeRunnable.run(SafeRunnable.java:176)
> 	at org.eclipse.jface.viewers.CellEditor.fireApplyEditorValue(CellEditor.java:332)
> 	at org.eclipse.jface.viewers.TextCellEditor.handleDefaultSelection(TextCellEditor.java:286)
> 	at org.eclipse.jface.viewers.TextCellEditor.lambda$0(TextCellEditor.java:146)
> 	at org.eclipse.swt.events.SelectionListener$2.widgetDefaultSelected(SelectionListener.java:102)
> 	at org.eclipse.swt.widgets.TypedListener.handleEvent(TypedListener.java:123)
> 	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
> 	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:5874)
> 	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1400)
> 	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:5138)
> 	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:4663)
> 	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1173)
> 	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
> 	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1062)
> 	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155)
> 	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:635)
> 	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
> 	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:559)
> 	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:150)
> 	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:155)
> 	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)
> 	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:137)
> 	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:107)
> 	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:400)
> 	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
> 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> 	at java.lang.reflect.Method.invoke(Method.java:498)
> 	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:660)
> 	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:597)
> 	at org.eclipse.equinox.launcher.Main.run(Main.java:1468)
> 	at org.eclipse.equinox.launcher.Main.main(Main.java:1441)
> Caused by: java.lang.ClassCastException: MWC.TacticalData.NarrativeEntry cannot be cast to MWC.GUI.Properties.PlainPropertyEditor$PropertyChangeAction
> 	at MWC.TacticalData.NarrativeWrapper$1.propertyChange(NarrativeWrapper.java:431)
> 	at java.beans.PropertyChangeSupport.fire(PropertyChangeSupport.java:335)
> 	at java.beans.PropertyChangeSupport.firePropertyChange(PropertyChangeSupport.java:328)
> 	at MWC.GUI.Editable$EditorType.fireChanged(Editable.java:968)
> 	at org.mwc.cmap.core.property_support.EditableWrapper$PropertyChangeAction.execute(EditableWrapper.java:162)
> 	at org.eclipse.core.commands.operations.DefaultOperationHistory.execute(DefaultOperationHistory.java:495)
> 	... 43 more
> org.eclipse.core.commands.ExecutionException: While executing the operation, an exception occurred
> 	at org.eclipse.core.commands.operations.DefaultOperationHistory.execute(DefaultOperationHistory.java:503)
> 	at org.mwc.cmap.core.CorePlugin.run(CorePlugin.java:823)
> 	at org.mwc.cmap.core.property_support.EditableWrapper.setPropertyValue(EditableWrapper.java:772)
> 	at org.eclipse.ui.views.properties.PropertySheetEntry.valueChanged(PropertySheetEntry.java:751)
> 	at org.eclipse.ui.views.properties.PropertySheetEntry.valueChanged(PropertySheetEntry.java:756)
> 	at org.eclipse.ui.views.properties.PropertySheetEntry.setValue(PropertySheetEntry.java:691)
> 	at org.eclipse.ui.views.properties.PropertySheetEntry.applyEditorValue(PropertySheetEntry.java:153)
> 	at org.eclipse.ui.views.properties.PropertySheetEntry$1.applyEditorValue(PropertySheetEntry.java:114)
> 	at org.eclipse.jface.viewers.CellEditor$1.run(CellEditor.java:335)
> 	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:45)
> 	at org.eclipse.ui.internal.JFaceUtil.lambda$0(JFaceUtil.java:47)
> 	at org.eclipse.jface.util.SafeRunnable.run(SafeRunnable.java:176)
> 	at org.eclipse.jface.viewers.CellEditor.fireApplyEditorValue(CellEditor.java:332)
> 	at org.eclipse.jface.viewers.TextCellEditor.handleDefaultSelection(TextCellEditor.java:286)
> 	at org.eclipse.jface.viewers.TextCellEditor.lambda$0(TextCellEditor.java:146)
> 	at org.eclipse.swt.events.SelectionListener$2.widgetDefaultSelected(SelectionListener.java:102)
> 	at org.eclipse.swt.widgets.TypedListener.handleEvent(TypedListener.java:123)
> 	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
> 	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:5874)
> 	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1400)
> 	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:5138)
> 	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:4663)
> 	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1173)
> 	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
> 	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1062)
> 	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155)
> 	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:635)
> 	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
> 	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:559)
> 	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:150)
> 	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:155)
> 	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)
> 	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:137)
> 	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:107)
> 	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:400)
> 	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
> 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> 	at java.lang.reflect.Method.invoke(Method.java:498)
> 	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:660)
> 	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:597)
> 	at org.eclipse.equinox.launcher.Main.run(Main.java:1468)
> 	at org.eclipse.equinox.launcher.Main.main(Main.java:1441)
> Caused by: java.lang.ClassCastException: MWC.TacticalData.NarrativeEntry cannot be cast to MWC.GUI.Properties.PlainPropertyEditor$PropertyChangeAction
> 	at MWC.TacticalData.NarrativeWrapper$1.propertyChange(NarrativeWrapper.java:431)
> 	at java.beans.PropertyChangeSupport.fire(PropertyChangeSupport.java:335)
> 	at java.beans.PropertyChangeSupport.firePropertyChange(PropertyChangeSupport.java:328)
> 	at MWC.GUI.Editable$EditorType.fireChanged(Editable.java:968)
> 	at org.mwc.cmap.core.property_support.EditableWrapper$PropertyChangeAction.execute(EditableWrapper.java:162)
> 	at org.eclipse.core.commands.operations.DefaultOperationHistory.execute(DefaultOperationHistory.java:495)
> 	... 43 more
>

--- a/org.mwc.cmap.core/src/org/mwc/cmap/core/property_support/DTGHelper.java
+++ b/org.mwc.cmap.core/src/org/mwc/cmap/core/property_support/DTGHelper.java
@@ -30,6 +30,7 @@ import org.eclipse.ui.views.properties.TextPropertyDescriptor;
 import org.mwc.cmap.core.CorePlugin;
 
 import MWC.GenericData.HiResDate;
+import MWC.TacticalData.NarrativeWrapper.GetHiResValue;
 import MWC.Utilities.TextFormatting.GMTDateFormat;
 
 public class DTGHelper extends EditorHelper
@@ -67,7 +68,8 @@ public class DTGHelper extends EditorHelper
     }
   }
 
-  public static class DTGPropertySource implements IPropertySource2
+  public static class DTGPropertySource implements IPropertySource2,
+  GetHiResValue
   {
 
     protected String _date, _time;

--- a/org.mwc.cmap.core/src/org/mwc/cmap/core/property_support/DTGHelper.java
+++ b/org.mwc.cmap.core/src/org/mwc/cmap/core/property_support/DTGHelper.java
@@ -10,7 +10,7 @@
  *
  *    This library is distributed in the hope that it will be useful,
  *    but WITHOUT ANY WARRANTY; without even the implied warranty of
- *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. 
+ *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
  */
 package org.mwc.cmap.core.property_support;
 
@@ -18,7 +18,7 @@ import java.text.ParseException;
 import java.text.SimpleDateFormat;
 import java.util.Date;
 
-import org.eclipse.core.runtime.Status;
+import org.eclipse.core.runtime.IStatus;
 import org.eclipse.jface.viewers.CellEditor;
 import org.eclipse.jface.viewers.ILabelProvider;
 import org.eclipse.jface.viewers.LabelProvider;
@@ -36,56 +36,18 @@ import MWC.Utilities.TextFormatting.GMTDateFormat;
 public class DTGHelper extends EditorHelper
 {
 
-  protected static SimpleDateFormat _longDateFormat;
-
-  protected static SimpleDateFormat _shortDateFormat;
-
-  protected static SimpleDateFormat _longTimeFormat;
-
-  protected static SimpleDateFormat _shortTimeFormat;
-
-  protected static SimpleDateFormat _fullFormat;
-
-  protected final static String LONG_DATE_FORMAT_DEFN = "dd/MMM/yyyy";
-  protected final static String SHORT_DATE_FORMAT_DEFN = "dd/MM/yyyy";
-
-  protected final static String LONG_TIME_FORMAT_DEFN = "HH:mm:ss";
-
-  protected final static String SHORT_TIME_FORMAT_DEFN = "HH:mm";
-
-  protected final static String UNSET = "unset";
-
-  protected synchronized static void checkDateFormat()
-  {
-    if (_longDateFormat == null)
-    {
-      _shortDateFormat = new GMTDateFormat(SHORT_DATE_FORMAT_DEFN);
-      _longDateFormat = new GMTDateFormat(LONG_DATE_FORMAT_DEFN);
-      _longTimeFormat = new GMTDateFormat(LONG_TIME_FORMAT_DEFN);
-      _shortTimeFormat = new GMTDateFormat(SHORT_TIME_FORMAT_DEFN);
-      _fullFormat = new GMTDateFormat(LONG_DATE_FORMAT_DEFN + "Z"
-          + LONG_TIME_FORMAT_DEFN);
-    }
-  }
-
   public static class DTGPropertySource implements IPropertySource2,
-  GetHiResValue
+      GetHiResValue
   {
-
-    protected String _date, _time;
-    protected String _originalDate, _originalTime;
-
-    protected HiResDate _originalVal;
 
     /**
      * name for the date property
-     * 
+     *
      */
     public static String ID_DATE = "Date";
-
     /**
      * name for the time property
-     * 
+     *
      */
     public static String ID_TIME = "Time";
 
@@ -97,6 +59,12 @@ public class DTGHelper extends EditorHelper
       {new TextPropertyDescriptor(ID_DATE, "date (dd/mmm/yyyy)"),
           new TextPropertyDescriptor(ID_TIME, "time (hh:mm:ss)"),};
     }
+
+    protected String _date, _time;
+
+    protected String _originalDate, _originalTime;
+
+    protected HiResDate _originalVal;
 
     public DTGPropertySource(final HiResDate dtg)
     {
@@ -121,13 +89,13 @@ public class DTGHelper extends EditorHelper
     }
 
     @Override
-    public boolean equals(Object obj)
+    public boolean equals(final Object obj)
     {
       final boolean res;
 
       if (obj instanceof DTGPropertySource)
       {
-        DTGPropertySource o = (DTGPropertySource) obj;
+        final DTGPropertySource o = (DTGPropertySource) obj;
         res = this.getValue().compareTo(o.getValue()) == 0;
       }
       else
@@ -142,16 +110,19 @@ public class DTGHelper extends EditorHelper
     {
     }
 
+    @Override
     public Object getEditableValue()
     {
       return this;
     }
 
+    @Override
     public IPropertyDescriptor[] getPropertyDescriptors()
     {
       return descriptors;
     }
 
+    @Override
     public Object getPropertyValue(final Object propName)
     {
       final String res;
@@ -169,10 +140,11 @@ public class DTGHelper extends EditorHelper
         throw new IllegalArgumentException(
             "We're not expecting property titled:" + propName);
       }
-      
+
       return res;
     }
 
+    @Override
     public synchronized HiResDate getValue()
     {
       HiResDate res = _originalVal;
@@ -188,7 +160,7 @@ public class DTGHelper extends EditorHelper
             final Date date = _longDateFormat.parse(_date);
             millis += date.getTime();
           }
-          catch (ParseException pe)
+          catch (final ParseException pe)
           {
             final Date date = _shortDateFormat.parse(_date);
             millis += date.getTime();
@@ -221,16 +193,24 @@ public class DTGHelper extends EditorHelper
       catch (final ParseException e)
       {
         // fall back on the original value
-        CorePlugin.logError(Status.ERROR,
+        CorePlugin.logError(IStatus.ERROR,
             "Failed to produce dtg from date/time strings", e);
         res = _originalVal;
       }
       return res;
     }
 
+    @Override
+    public boolean isPropertyResettable(final Object id)
+    {
+      // both parameters are resettable. cool.
+      return true;
+    }
+
     /**
      * @see org.eclipse.ui.views.properties.IPropertySource#isPropertySet(Object)
      */
+    @Override
     public boolean isPropertySet(final Object propName)
     {
       // always return true; so we can offer "reset" behaviour
@@ -247,6 +227,7 @@ public class DTGHelper extends EditorHelper
       return res;
     }
 
+    @Override
     public void resetPropertyValue(final Object propName)
     {
       if (ID_DATE.equals(propName))
@@ -259,6 +240,7 @@ public class DTGHelper extends EditorHelper
       }
     }
 
+    @Override
     public void setPropertyValue(final Object propName, final Object value)
     {
       if (ID_DATE.equals(propName))
@@ -272,6 +254,7 @@ public class DTGHelper extends EditorHelper
       firePropertyChanged((String) propName);
     }
 
+    @Override
     public String toString()
     {
       String res;
@@ -286,12 +269,38 @@ public class DTGHelper extends EditorHelper
       return res;
     }
 
-    public boolean isPropertyResettable(final Object id)
-    {
-      // both parameters are resettable. cool.
-      return true;
-    }
+  }
 
+  protected static SimpleDateFormat _longDateFormat;
+
+  protected static SimpleDateFormat _shortDateFormat;
+
+  protected static SimpleDateFormat _longTimeFormat;
+
+  protected static SimpleDateFormat _shortTimeFormat;
+
+  protected static SimpleDateFormat _fullFormat;
+  protected final static String LONG_DATE_FORMAT_DEFN = "dd/MMM/yyyy";
+
+  protected final static String SHORT_DATE_FORMAT_DEFN = "dd/MM/yyyy";
+
+  protected final static String LONG_TIME_FORMAT_DEFN = "HH:mm:ss";
+
+  protected final static String SHORT_TIME_FORMAT_DEFN = "HH:mm";
+
+  protected final static String UNSET = "unset";
+
+  protected synchronized static void checkDateFormat()
+  {
+    if (_longDateFormat == null)
+    {
+      _shortDateFormat = new GMTDateFormat(SHORT_DATE_FORMAT_DEFN);
+      _longDateFormat = new GMTDateFormat(LONG_DATE_FORMAT_DEFN);
+      _longTimeFormat = new GMTDateFormat(LONG_TIME_FORMAT_DEFN);
+      _shortTimeFormat = new GMTDateFormat(SHORT_TIME_FORMAT_DEFN);
+      _fullFormat = new GMTDateFormat(LONG_DATE_FORMAT_DEFN + "Z"
+          + LONG_TIME_FORMAT_DEFN);
+    }
   }
 
   public DTGHelper()
@@ -299,11 +308,7 @@ public class DTGHelper extends EditorHelper
     super(HiResDate.class);
   }
 
-  public CellEditor getCellEditorFor(final Composite parent)
-  {
-    return null;
-  }
-
+  @Override
   @SuppressWarnings(
   {"rawtypes"})
   public boolean editsThis(final Class target)
@@ -311,23 +316,24 @@ public class DTGHelper extends EditorHelper
     return (target == HiResDate.class);
   }
 
-  public Object translateToSWT(final Object value)
+  @Override
+  public CellEditor getCellEditorFor(final Composite parent)
   {
-    // ok, we've received a DTG. Return our new property source representing a
-    // DTG
-    return new DTGPropertySource((HiResDate) value);
+    return null;
   }
 
-  public Object translateFromSWT(final Object value)
-  {
-    final DTGPropertySource res = (DTGPropertySource) value;
-    return res.getValue();
-  }
-
+  @Override
   public ILabelProvider getLabelFor(final Object currentValue)
   {
     final ILabelProvider label1 = new LabelProvider()
     {
+      @Override
+      public Image getImage(final Object element)
+      {
+        return null;
+      }
+
+      @Override
       public String getText(final Object element)
       {
         final DTGPropertySource val = (DTGPropertySource) element;
@@ -344,12 +350,22 @@ public class DTGHelper extends EditorHelper
         return res;
       }
 
-      public Image getImage(final Object element)
-      {
-        return null;
-      }
-
     };
     return label1;
+  }
+
+  @Override
+  public Object translateFromSWT(final Object value)
+  {
+    final DTGPropertySource res = (DTGPropertySource) value;
+    return res.getValue();
+  }
+
+  @Override
+  public Object translateToSWT(final Object value)
+  {
+    // ok, we've received a DTG. Return our new property source representing a
+    // DTG
+    return new DTGPropertySource((HiResDate) value);
   }
 }

--- a/org.mwc.cmap.legacy/src/MWC/TacticalData/NarrativeWrapper.java
+++ b/org.mwc.cmap.legacy/src/MWC/TacticalData/NarrativeWrapper.java
@@ -133,7 +133,6 @@ import MWC.GUI.Layer;
 import MWC.GUI.Layers;
 import MWC.GUI.Plottable;
 import MWC.GUI.TimeStampedDataItem;
-import MWC.GUI.Properties.PlainPropertyEditor.PropertyChangeAction;
 import MWC.GenericData.HiResDate;
 import MWC.GenericData.TimePeriod;
 import MWC.Utilities.ReaderWriter.ImportManager;
@@ -146,6 +145,11 @@ import junit.framework.TestCase;
 public final class NarrativeWrapper extends MWC.GUI.PlainWrapper implements
     Layer, IRollingNarrativeProvider, GriddableSeriesMarker
 {
+  
+  public static interface GetHiResValue
+  {
+    public HiResDate getValue();
+  }
 
   /*
    * embedded class to allow us to pass the local iterator (Iterator) used internally outside as an
@@ -427,14 +431,33 @@ public final class NarrativeWrapper extends MWC.GUI.PlainWrapper implements
         // double-check it's the date
         if (evt.getPropertyName().equals(NarrativeEntry.DTG))
         {
-          // ok, remove this entry
-          final PropertyChangeAction pcs = (PropertyChangeAction) evt
-              .getSource();
-          final NarrativeEntry entry = (NarrativeEntry) pcs.getData();
-          _myEntries.remove(entry);
-
-          // and replace it (which re-sort them)
-          _myEntries.add((Editable) entry);
+          if ( evt.getSource() instanceof NarrativeEntry )
+          {
+            final NarrativeEntry entry = (NarrativeEntry) evt.getSource();
+            
+            final Object oldValue = evt.getOldValue();
+            final Object newValue = evt.getNewValue();
+            if ( oldValue instanceof GetHiResValue && newValue instanceof GetHiResValue )
+            {
+              final HiResDate oldHiResValue = ((GetHiResValue)oldValue).getValue();
+              final HiResDate newHiResValue = ((GetHiResValue)newValue).getValue();
+              
+              /**
+               * This part was tricky. Previously we were editing the key,
+               * so the element was getting lost in the tree. So, what I am 
+               * doing here is to store the old value again to find the item
+               * in the tree, removing it, and then adding it again. The idea
+               * is to resort the items.
+               * 
+               * Saul
+               */
+              
+              entry.setDTG(oldHiResValue);
+              _myEntries.remove(entry);              
+              entry.setDTG(newHiResValue);
+              _myEntries.add((Editable) entry);
+            }
+          }
         }
       }
     };

--- a/org.mwc.cmap.legacy/src/MWC/TacticalData/NarrativeWrapper.java
+++ b/org.mwc.cmap.legacy/src/MWC/TacticalData/NarrativeWrapper.java
@@ -145,7 +145,7 @@ import junit.framework.TestCase;
 public final class NarrativeWrapper extends MWC.GUI.PlainWrapper implements
     Layer, IRollingNarrativeProvider, GriddableSeriesMarker
 {
-  
+
   public static interface GetHiResValue
   {
     public HiResDate getValue();
@@ -155,8 +155,7 @@ public final class NarrativeWrapper extends MWC.GUI.PlainWrapper implements
    * embedded class to allow us to pass the local iterator (Iterator) used internally outside as an
    * Enumeration
    */
-  protected static final class IteratorWrapper implements
-      Enumeration<Editable>
+  protected static final class IteratorWrapper implements Enumeration<Editable>
   {
     private final Iterator<Editable> _val;
 
@@ -196,19 +195,19 @@ public final class NarrativeWrapper extends MWC.GUI.PlainWrapper implements
       super(data, data.getName(), "Narrative");
     }
 
-//    /**
-//     * return a description of this bean, also specifies the custom editor we use
-//     *
-//     * @return the BeanDescriptor
-//     */
-//    @Override
-//    public final BeanDescriptor getBeanDescriptor()
-//    {
-//      final BeanDescriptor bp = new BeanDescriptor(NarrativeWrapper.class,
-//          Debrief.GUI.Panels.NarrativeViewer.class);
-//      bp.setDisplayName("Narrative Viewer");
-//      return bp;
-//    }
+    // /**
+    // * return a description of this bean, also specifies the custom editor we use
+    // *
+    // * @return the BeanDescriptor
+    // */
+    // @Override
+    // public final BeanDescriptor getBeanDescriptor()
+    // {
+    // final BeanDescriptor bp = new BeanDescriptor(NarrativeWrapper.class,
+    // Debrief.GUI.Panels.NarrativeViewer.class);
+    // bp.setDisplayName("Narrative Viewer");
+    // return bp;
+    // }
 
     @Override
     public final MethodDescriptor[] getMethodDescriptors()
@@ -431,31 +430,33 @@ public final class NarrativeWrapper extends MWC.GUI.PlainWrapper implements
         // double-check it's the date
         if (evt.getPropertyName().equals(NarrativeEntry.DTG))
         {
-          if ( evt.getSource() instanceof NarrativeEntry )
+          if (evt.getSource() instanceof NarrativeEntry)
           {
             final NarrativeEntry entry = (NarrativeEntry) evt.getSource();
-            
+
             final Object oldValue = evt.getOldValue();
             final Object newValue = evt.getNewValue();
-            if ( oldValue instanceof GetHiResValue && newValue instanceof GetHiResValue )
+            if (oldValue instanceof GetHiResValue
+                && newValue instanceof GetHiResValue)
             {
-              final HiResDate oldHiResValue = ((GetHiResValue)oldValue).getValue();
-              final HiResDate newHiResValue = ((GetHiResValue)newValue).getValue();
-              
+              final HiResDate oldHiResValue = ((GetHiResValue) oldValue)
+                  .getValue();
+              final HiResDate newHiResValue = ((GetHiResValue) newValue)
+                  .getValue();
+
               /**
-               * This part was tricky. Previously we were editing the key,
-               * so the element was getting lost in the tree. So, what I am 
-               * doing here is to store the old value again to find the item
-               * in the tree, removing it, and then adding it again. The idea
-               * is to resort the items.
-               * 
+               * This part was tricky. Previously we were editing the key, so the element was
+               * getting lost in the tree. So, what I am doing here is to store the old value again
+               * to find the item in the tree, removing it, and then adding it again. The idea is to
+               * resort the items.
+               *
                * Saul
                */
-              
+
               entry.setDTG(oldHiResValue);
-              _myEntries.remove(entry);              
+              _myEntries.remove(entry);
               entry.setDTG(newHiResValue);
-              _myEntries.add((Editable) entry);
+              _myEntries.add(entry);
             }
           }
         }

